File: src-tauri/src/protocol.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;

pub const PROTOCOL_VERSION: u8 = 1;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(tag = "type")]
pub enum RpcMessage {
    #[serde(rename = "req")]
    Request {
        v: u8,
        id: String,
        method: String,
        #[serde(skip_serializing_if = "Option::is_none")]
        params: Option<Value>,
    },
    #[serde(rename = "res")]
    Response {
        v: u8,
        id: String,
        #[serde(skip_serializing_if = "Option::is_none")]
        result: Option<Value>,
    },
    #[serde(rename = "err")]
    Error {
        v: u8,
        id: String,
        error: RpcErrorBody,
    },
    #[serde(rename = "evt")]
    Event {
        v: u8,
        method: String,
        #[serde(skip_serializing_if = "Option::is_none")]
        params: Option<Value>,
    },
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct RpcErrorBody {
    pub code: String,
    pub message: String,
}

pub fn encode_message(message: &RpcMessage) -> Result<String, serde_json::Error> {
    Ok(format!("{}\n", serde_json::to_string(message)?))
}

pub fn decode_message(line: &str) -> Result<RpcMessage, String> {
    let trimmed = line.trim();
    if trimmed.is_empty() {
        return Err("rpc: empty line".into());
    }
    let message: RpcMessage =
        serde_json::from_str(trimmed).map_err(|error| format!("rpc: {error}"))?;
    match &message {
        RpcMessage::Request { v, .. }
        | RpcMessage::Response { v, .. }
        | RpcMessage::Error { v, .. }
        | RpcMessage::Event { v, .. } => {
            if *v != PROTOCOL_VERSION {
                return Err("rpc: unsupported protocol version".into());
            }
        }
    }
    Ok(message)
}

pub fn ok_result(id: impl Into<String>, result: Value) -> RpcMessage {
    RpcMessage::Response {
        v: PROTOCOL_VERSION,
        id: id.into(),
        result: Some(result),
    }
}

pub fn err_result(id: impl Into<String>, message: impl Into<String>) -> RpcMessage {
    RpcMessage::Error {
        v: PROTOCOL_VERSION,
        id: id.into(),
        error: RpcErrorBody {
            code: "native".into(),
            message: message.into(),
        },
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn round_trips_request() {
        let encoded = encode_message(&RpcMessage::Request {
            v: 1,
            id: "1".into(),
            method: "shell.mount".into(),
            params: Some(serde_json::json!({"generationId":"g1"})),
        })
        .unwrap();
        assert!(encoded.ends_with('\n'));
        let decoded = decode_message(&encoded).unwrap();
        assert_eq!(
            decoded,
            RpcMessage::Request {
                v: 1,
                id: "1".into(),
                method: "shell.mount".into(),
                params: Some(serde_json::json!({"generationId":"g1"})),
            }
        );
    }

    #[test]
    fn rejects_foreign_version() {
        let error = decode_message(r#"{"type":"req","v":2,"id":"1","method":"shell.show"}"#).unwrap_err();
        assert!(error.contains("unsupported protocol version"));
    }
}

File: host/src/rpc/protocol.ts
export const PROTOCOL_VERSION = 1 as const

export type RpcRequest = {
  readonly v: typeof PROTOCOL_VERSION
  readonly type: 'req'
  readonly id: string
  readonly method: string
  readonly params?: unknown
}

export type RpcResponse = {
  readonly v: typeof PROTOCOL_VERSION
  readonly type: 'res'
  readonly id: string
  readonly result?: unknown
}

export type RpcError = {
  readonly v: typeof PROTOCOL_VERSION
  readonly type: 'err'
  readonly id: string
  readonly error: { readonly code: string, readonly message: string }
}

export type RpcEvent = {
  readonly v: typeof PROTOCOL_VERSION
  readonly type: 'evt'
  readonly method: string
  readonly params?: unknown
}

export type RpcMessage = RpcRequest | RpcResponse | RpcError | RpcEvent

export const NATIVE_METHODS = {
  schedule: 'shell.schedule',
  mount: 'shell.mount',
  release: 'shell.release',
  show: 'shell.show',
  notifyAttention: 'shell.notifyAttention',
  reloadRenderer: 'shell.reloadRenderer',
  toggleDeveloperTools: 'shell.toggleDeveloperTools',
  pickDirectory: 'shell.pickDirectory',
  validateDirectory: 'shell.validateDirectory',
  openDialog: 'shell.openDialog',
  openRecovery: 'shell.openRecovery',
  openProfileCreate: 'shell.openProfileCreate',
  openTerminal: 'shell.openTerminal',
  exportDiagnostics: 'shell.exportDiagnostics',
  setTheme: 'shell.setTheme',
  setLocale: 'shell.setLocale',
  trayUpsert: 'shell.tray.upsert',
  trayRemove: 'shell.tray.remove',
  notify: 'shell.notify',
  confirmUpdate: 'shell.confirmUpdate',
  showUpdateResult: 'shell.showUpdateResult',
  downloadUpdate: 'shell.downloadUpdate',
  restart: 'shell.restart',
  restartRecovery: 'shell.restartRecovery',
  prepareToQuit: 'shell.prepareToQuit',
  reportRendererBoot: 'shell.reportRendererBoot',
  writeBootstrap: 'shell.writeBootstrap',
  openDevtools: 'shell.openDevtools',
  revealItem: 'shell.revealItem',
  openUpdate: 'shell.openUpdate',
  saveDialog: 'shell.saveDialog',
} as const

export const NATIVE_EVENTS = {
  secondInstance: 'event.secondInstance',
  trayInvoke: 'event.trayInvoke',
  dropPaths: 'event.dropPaths',
  windowClose: 'event.windowClose',
  quit: 'event.quit',
  sidecarFailed: 'event.sidecarFailed',
} as const

export function encodeMessage(message: RpcMessage): string {
  return `${JSON.stringify(message)}\n`
}

export function decodeMessage(line: string): RpcMessage {
  const trimmed = line.trim()
  if (trimmed.length === 0) throw new Error('rpc: empty line')
  const value: unknown = JSON.parse(trimmed)
  if (value === null || typeof value !== 'object' || Array.isArray(value)) {
    throw new Error('rpc: message must be an object')
  }
  const record = value as Record<string, unknown>
  if (record.v !== PROTOCOL_VERSION) throw new Error('rpc: unsupported protocol version')
  if (record.type === 'req') {
    if (typeof record.id !== 'string' || typeof record.method !== 'string') {
      throw new Error('rpc: invalid request')
    }
    return {
      v: PROTOCOL_VERSION,
      type: 'req',
      id: record.id,
      method: record.method,
      ...(record.params === undefined ? {} : { params: record.params }),
    }
  }
  if (record.type === 'res') {
    if (typeof record.id !== 'string') throw new Error('rpc: invalid response')
    return {
      v: PROTOCOL_VERSION,
      type: 'res',
      id: record.id,
      ...(record.result === undefined ? {} : { result: record.result }),
    }
  }
  if (record.type === 'err') {
    if (typeof record.id !== 'string' || record.error === null || typeof record.error !== 'object') {
      throw new Error('rpc: invalid error')
    }
    const error = record.error as Record<string, unknown>
    if (typeof error.code !== 'string' || typeof error.message !== 'string') {
      throw new Error('rpc: invalid error payload')
    }
    return {
      v: PROTOCOL_VERSION,
      type: 'err',
      id: record.id,
      error: { code: error.code, message: error.message },
    }
  }
  if (record.type === 'evt') {
    if (typeof record.method !== 'string') throw new Error('rpc: invalid event')
    return {
      v: PROTOCOL_VERSION,
      type: 'evt',
      method: record.method,
      ...(record.params === undefined ? {} : { params: record.params }),
    }
  }
  throw new Error('rpc: unknown message type')
}
